File: android/app/src/main/java/com/vixel/studio/core/store/ProjectStore.kt
package com.vixel.studio.core.store

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.util.Log
import com.vixel.studio.core.model.Project
import org.json.JSONObject
import java.io.File

private const val TAG = "VixelProjects"

data class ProjectSummary(
    val id: String,
    val name: String,
    val durationUs: Long,
    val clipCount: Int,
    val overlayCount: Int,
    val aspectId: String,
    val updatedAt: Long,
)

/**
 * Projects on disk, one JSON file each under `files/projects`.
 *
 * Media is referenced by uri rather than copied. Copying would make a project
 * self-contained but would also duplicate every video the user edits, so the
 * tradeoff here is storage in exchange for [missingMedia] being possible —
 * see its note.
 */
object ProjectStore {

    private const val EXTENSION = ".vixel.json"

    fun directory(context: Context): File =
        File(context.filesDir, "projects").apply { mkdirs() }

    private fun fileFor(context: Context, id: String): File =
        File(directory(context), "$id$EXTENSION")

    fun save(context: Context, project: Project): Boolean = try {
        val json = ProjectJson.toJson(project).toString()
        val target = fileFor(context, project.id)
        // Write beside the target and swap, so an interrupted save cannot
        // leave a half-written project that fails to open.
        val temp = File(target.parentFile, "${target.name}.tmp")
        temp.writeText(json)
        if (target.exists()) target.delete()
        val renamed = temp.renameTo(target)
        if (!renamed) temp.copyTo(target, overwrite = true).also { temp.delete() }
        true
    } catch (t: Throwable) {
        Log.e(TAG, "could not save project ${project.id}", t)
        false
    }

    fun load(context: Context, id: String): Project? = try {
        val file = fileFor(context, id)
        if (!file.exists()) {
            null
        } else {
            ProjectJson.fromJson(JSONObject(file.readText()))
        }
    } catch (t: Throwable) {
        Log.e(TAG, "could not load project $id", t)
        null
    }

    /** Newest first. */
    fun list(context: Context): List<ProjectSummary> =
        directory(context)
            .listFiles { file -> file.isFile && file.name.endsWith(EXTENSION) }
            ?.mapNotNull { file -> summarise(file) }
            ?.sortedByDescending { it.updatedAt }
            ?: emptyList()

    private fun summarise(file: File): ProjectSummary? = try {
        val project = ProjectJson.fromJson(JSONObject(file.readText()))
        ProjectSummary(
            id = project.id,
            name = project.name,
            durationUs = project.durationUs,
            clipCount = project.clips.size,
            overlayCount = project.overlays.size,
            aspectId = project.aspect.id,
            updatedAt = file.lastModified(),
        )
    } catch (t: Throwable) {
        Log.w(TAG, "skipping unreadable project ${file.name}", t)
        null
    }

    fun delete(context: Context, id: String): Boolean = fileFor(context, id).delete()

    fun exists(context: Context, id: String): Boolean = fileFor(context, id).exists()

    /**
     * Clips whose media can no longer be opened.
     *
     * A picked uri is a grant, not a copy, and a grant can lapse — the file is
     * deleted, the SD card is pulled, or the permission expires. Reporting this
     * lets the editor say which clip is unavailable instead of rendering black
     * frames and leaving the user to guess.
     */
    fun missingMedia(context: Context, project: Project): List<String> =
        project.clips.filterNot { canRead(context, it.uri) }.map { it.id }

    fun canRead(context: Context, uri: String): Boolean = try {
        context.contentResolver.openInputStream(Uri.parse(uri))?.use { true } ?: false
    } catch (t: Throwable) {
        false
    }

    /**
     * Asks for long-lived access to a picked uri.
     *
     * Only document-provider uris can grant this; the photo picker cannot, and
     * throws. Failing is normal, so it is swallowed.
     */
    fun tryPersistPermission(context: Context, uri: Uri) {
        try {
            context.contentResolver.takePersistableUriPermission(
                uri,
                Intent.FLAG_GRANT_READ_URI_PERMISSION,
            )
        } catch (t: Throwable) {
            Log.d(TAG, "uri is not persistable, continuing with the session grant")
        }
    }
}

File: flutter/lib/core/store/project_store.dart
import 'dart:convert';
import 'dart:io';

import 'package:path_provider/path_provider.dart';

import '../model/project.dart';

/// Summary of a stored project, for listing without parsing every file.
class ProjectSummary {
  const ProjectSummary({
    required this.id,
    required this.name,
    required this.clipCount,
    required this.durationUs,
    required this.updatedAt,
  });

  final String id;
  final String name;
  final int clipCount;
  final int durationUs;
  final DateTime updatedAt;
}

/// Projects on disk, one JSON file each.
///
/// A file per project rather than one index: a corrupt write can then only
/// cost the project being edited, not the whole library, and listing is a
/// directory read instead of a parse of everything.
class ProjectStore {
  const ProjectStore._();

  static const String _folder = 'projects';

  static Future<Directory> _dir() async {
    final base = await getApplicationDocumentsDirectory();
    final dir = Directory('${base.path}/$_folder');
    if (!await dir.exists()) await dir.create(recursive: true);
    return dir;
  }

  static Future<File> _fileFor(String id) async =>
      File('${(await _dir()).path}/$id.json');

  /// Writes [project], via a temporary file that is renamed into place.
  ///
  /// Autosave fires while the user is still editing, so a write interrupted by
  /// the process being killed has to leave the previous save intact rather
  /// than a half-written file where the project used to be. Rename is atomic;
  /// writing in place is not.
  static Future<void> save(Project project) async {
    final file = await _fileFor(project.id);
    final temp = File('${file.path}.tmp');
    await temp.writeAsString(jsonEncode(project.toJson()), flush: true);
    await temp.rename(file.path);
  }

  static Future<Project?> load(String id) async {
    try {
      final file = await _fileFor(id);
      if (!await file.exists()) return null;
      final json = jsonDecode(await file.readAsString());
      if (json is! Map) return null;
      return Project.fromJson(Map<String, dynamic>.from(json));
    } catch (_) {
      // A project that will not parse is reported as missing rather than
      // thrown from: the caller's only useful response is the same either way,
      // and a crash on the home screen would make the app unopenable.
      return null;
    }
  }

  static Future<List<ProjectSummary>> list() async {
    final dir = await _dir();
    final summaries = <ProjectSummary>[];

    await for (final entry in dir.list()) {
      if (entry is! File || !entry.path.endsWith('.json')) continue;
      try {
        final json = jsonDecode(await entry.readAsString());
        if (json is! Map) continue;
        final project = Project.fromJson(Map<String, dynamic>.from(json));
        summaries.add(
          ProjectSummary(
            id: project.id,
            name: project.name,
            clipCount: project.clips.length,
            durationUs: project.durationUs,
            updatedAt: await entry.lastModified(),
          ),
        );
      } catch (_) {
        continue;
      }
    }

    summaries.sort((a, b) => b.updatedAt.compareTo(a.updatedAt));
    return summaries;
  }

  static Future<void> delete(String id) async {
    final file = await _fileFor(id);
    if (await file.exists()) await file.delete();
  }

  /// Ids of clips whose media can no longer be opened.
  ///
  /// A picked file is a reference, not a copy, so it can be deleted from under
  /// a saved project. Naming the affected clips lets the editor say which ones
  /// are unavailable instead of rendering black frames and leaving the user to
  /// work out which of twenty clips went missing.
  static Future<Set<String>> missingMedia(Project project) async {
    final missing = <String>{};
    for (final clip in project.clips) {
      final path = Uri.tryParse(clip.uri)?.toFilePath(windows: false);
      if (path == null) continue;
      if (!await File(path).exists()) missing.add(clip.id);
    }
    return missing;
  }
}
